### ISSUE TITLE
[Bug]: DDP with epochCounterWithStopper leads to CUDA timeouts

### ISSUE THREAD

**Issue by matthewkperez** [speechbrain/speechbrain#1939]
### Describe the bug

I'm currently using DDP and EpochCounterWithStopper in order to speed up training. However, on my experiments I am running into timeout errors,

`Watchdog caught collective operation timeout: WorkNCCL(SeqNum=588758, OpType=ALLREDUCE, Timeout(ms)=3600000) ran for 3602438 milliseconds before timing out.`

From tinkering around with it, I think this is due to the fact that I am using DDP and where I have placed the epoch_counter.should_stop() as it seems to run this for each GPU and the timeout occurs when one GPU hits a early stop while the other does not.

### Expected behaviour

EpochCounterWithStopper to work on multiple GPUs using DDP.

### To Reproduce

Here is my fit function where I have the early stopping

I am currently running on 2 GPUS

`def fit(
        self,
        epoch_counter,
        train_set,
        valid_set=None,
        progressbar=None,
        train_loader_kwargs={},
        valid_loader_kwargs={},
    ):
        """Iterate epochs and datasets to improve objective.

        Relies on the existence of multiple functions that can (or should) be
        overridden. The following methods are used and expected to have a
        certain behavior:

        * ``fit_batch()``
        * ``evaluate_batch()``
        * ``update_average()``

        If the initialization was done with distributed_count > 0 and the
        distributed_backend is ddp, this will generally handle multiprocess
        logic, like splitting the training data into subsets for each device and
        only saving a checkpoint on the main process.

        Arguments
        ---------
        epoch_counter : iterable
            Each call should return an integer indicating the epoch count.
        train_set : Dataset, DataLoader
            A set of data to use for training. If a Dataset is given, a
            DataLoader is automatically created. If a DataLoader is given, it is
            used directly.
        valid_set : Dataset, DataLoader
            A set of data to use for validation. If a Dataset is given, a
            DataLoader is automatically created. If a DataLoader is given, it is
            used directly.
        train_loader_kwargs : dict
            Kwargs passed to `make_dataloader()` for making the train_loader
            (if train_set is a Dataset, not DataLoader).
            E.G. batch_size, num_workers.
            DataLoader kwargs are all valid.
        valid_loader_kwargs : dict
            Kwargs passed to `make_dataloader()` for making the valid_loader
            (if valid_set is a Dataset, not DataLoader).
            E.g., batch_size, num_workers.
            DataLoader kwargs are all valid.
        progressbar : bool
            Whether to display the progress of each epoch in a progressbar.
        """
        if not (
            isinstance(train_set, DataLoader)
            or isinstance(train_set, LoopedLoader)
        ):
            train_set = self.make_dataloader(
                train_set, stage=sb.Stage.TRAIN, **train_loader_kwargs
            )
        if valid_set is not None and not (
            isinstance(valid_set, DataLoader)
            or isinstance(valid_set, LoopedLoader)
        ):
            valid_set = self.make_dataloader(
                valid_set,
                stage=sb.Stage.VALID,
                ckpt_prefix=None,
                **valid_loader_kwargs,
            )

        self.on_fit_start()

        if progressbar is None:
            progressbar = not self.noprogressbar

        # Only show progressbar if requested and main_process
        enable = progressbar and sb.utils.distributed.if_main_process()

        # epoch_counter (load from checkpoint)
        if isinstance(epoch_counter, sb.utils.epoch_loop.EpochCounterWithStopper):
            if epoch_counter.should_stop(current=epoch_counter.current,current_metric=float('inf')):
                epoch_counter.current = epoch_counter.limit  # skipping unpromising epochs, set curr_epoch to last epoch
                print("early stopping proc")

        # Iterate epochs
        for epoch in epoch_counter:
            self._fit_train(train_set=train_set, epoch=epoch, enable=enable)
            self._fit_valid(valid_set=valid_set, epoch=epoch, enable=enable)

            # epoch_counter
            if isinstance(epoch_counter, sb.utils.epoch_loop.EpochCounterWithStopper):
                if epoch_counter.should_stop(current=epoch,current_metric=self.valid_loss):
                    epoch_counter.current = epoch_counter.limit  # skipping unpromising epochs, set curr_epoch to last epoch
                    print("early stopping proc")

            # Debug mode only runs a few epochs
            if (
                self.debug
                and epoch == self.debug_epochs
                or self._optimizer_step_limit_exceeded
            ):
                break
`

Epoch_Loop.py

`
class EpochCounterWithStopper(EpochCounter):
    """An epoch counter which can save and recall its state, integrating an early stopper by tracking a target metric.

    Arguments
    ---------
    limit: int
        maximum number of epochs
    limit_to_stop : int
        maximum number of consecutive epochs without improvements in performance
    limit_warmup : int
        number of epochs to wait until start checking for early stopping
    direction : "max" or "min"
        direction to optimize the target metric

    Example
    -------
    >>> limit = 10
    >>> limit_to_stop = 5
    >>> limit_warmup = 2
    >>> direction = "min"
    >>> epoch_counter = EpochCounterWithStopper(limit, limit_to_stop, limit_warmup, direction)
    >>> for epoch in epoch_counter:
    ...     # Run training...
    ...     # Track a validation metric,
    ...     current_valid_metric = 0
    ...     # get the current valid metric (get current_valid_metric)
    ...     if epoch_counter.should_stop(current=epoch,
    ...                                  current_metric=current_valid_metric,):
    ...         epoch_counter.current = epoch_counter.limit  # skipping unpromising epochs
    """

    def __init__(self, limit, limit_to_stop, limit_warmup, direction):
        super().__init__(limit)
        self.limit_to_stop = limit_to_stop
        self.limit_warmup = limit_warmup
        self.direction = direction

        self.best_limit = 0
        self.min_delta = 1e-6

        if self.limit_to_stop < 0:
            raise ValueError("Stopper 'limit_to_stop' must be >= 0")
        if self.limit_warmup < 0:
            raise ValueError("Stopper 'limit_warmup' must be >= 0")
        if self.direction == "min":
            self.th, self.sign = float("inf"), 1
        elif self.direction == "max":
            self.th, self.sign = -float("inf"), -1
        else:
            raise ValueError("Stopper 'direction' must be 'min' or 'max'")

    def should_stop(self, current, current_metric):
        """Returns True is training should stop (based on the performance
        metrics)."""
        should_stop = False
        print(f"Epoch Counter: cur-ep={current}, best-ep={self.best_limit}, limit={self.limit_to_stop}")
        if current > self.limit_warmup:
            if self.sign * current_metric < self.sign * (
                (1 - self.min_delta) * self.th
            ):
                self.best_limit = current
                self.th = current_metric
            should_stop = (current - self.best_limit) >= self.limit_to_stop
        return should_stop

    @mark_as_saver
    def _save(self, path):
        with open(path, "w") as fo:
            fo.write(f"current: {str(self.current)}\n")
            fo.write(f"best_epoch: {str(self.best_limit)}\n")
            fo.write(f"best_loss: {str(self.th)}\n")

    @mark_as_loader
    def _recover(self, path, end_of_epoch=True, device=None):
        # NOTE: end_of_epoch = True by default so that when
        #  loaded in parameter transfer, this starts a new epoch.
        #  However, parameter transfer to EpochCounter should
        #  probably never be used really.
        del device  # Not used.
        with open(path) as fi:
            # print(fi.readlines())
            current = int(fi.readline().split()[-1].strip())
            self.best_limit = int(fi.readline().split()[-1].strip())
            self.th = float(fi.readline().split()[-1].strip())
            if end_of_epoch:
                self.current = current
            else:
                self.current = current - 1
`

### Relevant log output

```shell
Epoch Counter: cur-ep=32, best-ep=29, limit=3
early stopping proc
Epoch Counter: cur-ep=32, best-ep=30, limit=3
speechbrain.utils.epoch_loop - Going into epoch 33
 84%|██████████████████████████████████████████████████████████████████████████████████████████▌                 | 2197/2620 [59:58<29:57,  4.25s/it][E ProcessGroupNCCL.cpp:821] [Rank 0] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=588758, OpType=ALLREDUCE, Timeout(ms)=3600000) ran for 3602438 milliseconds before timing out.
  0%|                                                                                                    | 0/1784 [1:00:22<?, ?it/s, train_loss=9.92][E ProcessGroupNCCL.cpp:456] Some NCCL operations have failed or timed out. Due to the asynchronous nature of CUDA kernels, subsequent GPU operations might run on corrupted/incomplete data.
[E ProcessGroupNCCL.cpp:461] To avoid data inconsistency, we are taking the entire process down.
terminate called after throwing an instance of 'std::runtime_error'
  what():  [Rank 0] Watchdog caught collective operation timeout: WorkNCCL(SeqNum=588758, OpType=ALLREDUCE, Timeout(ms)=3600000) ran for 3602438 milliseconds before timing out.
```

### Additional context

I have added print statements in the epoch_loop.should_stop() to log the epoch counter variables. Based on the output logs, the epoch counter displays its characteristics twice for each epoch, usually these are the same and based on my understanding of DDP, I would expect the epoch counter to be exactly the same as it is training. However, when it errors, the epoch counter for 1 GPU seems to have detected an early stopping while the other is continuing to train.

When googling the timeout error, the torch "fix" seems to be just to increase the timeout parameter but I dont think this is the issue in my case.

**Comment by My-captain** [comment]
I got the same problem last night.

**Comment by Adel-Moumen** [comment]
Hello guys thanks for raising this issue.

@AsuMagic could you please have a look? Thanks!

**Comment by Adel-Moumen** [comment]
Hello @matthewkperez and @My-captain, 

Thanks for opening this issue. 

Could you please retry? The PR #2051 has been merged in the develop branch and should fixed some issues with our `EpochCounterWithStopper`. Thanks! 

Adel

**Comment by Adel-Moumen** [comment]
Hello, 

It's been quite some time without any updates or news on this issue. Consequently, I'm taking the step to close it. Feel free to reopen it if you are still facing the issue. 

Note: as explained, the PR #2051 should have fixed this issue. 

Best,
Adel
